File: baselines/omr/sender_smoke.cpp
#include "include/regevEncryption.h"

#include <chrono>
#include <algorithm>
#include <cstdlib>
#include <ctime>
#include <fstream>
#include <iostream>
#include <string>
#include <unistd.h>

using Clock = std::chrono::steady_clock;

static void append(const std::string &path, int trial, bool warm, const char *op,
                   long long elapsed, const char *bytes_status, const char *status) {
    bool header = !std::ifstream(path).good(); std::ofstream out(path, std::ios::app);
    if (header) out << "timestamp,scheme,experiment,operation,role,trial,warmup,message_size_bytes,recipient_input_size,scheme_specific_parameters,wall_time_ns,output_bytes,output_bytes_status,status,git_commit,compiler,compiler_version,build_type,machine_id\n";
    char host[256] = {}; gethostname(host, sizeof(host)-1);
    auto ts=std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
    out << ts << ",OMR,sender_signaling," << op << ",sender," << trial << ',' << (warm?"true":"false")
        << ",not_applicable,unclear,PVW_n=450;PVW_q=65537;PVW_stddev=1.3;PVW_m=16000;PVW_ell=4,"
        << elapsed << ",," << bytes_status << ',' << status << ",unavailable,GCC," << __VERSION__
        << ",Release," << host << '\n'; out.flush();
}

static void append_scale(const std::string &path,int B,int trial,bool warm,long long total,long long per,const char *status){
 bool header=!std::ifstream(path).good(); std::ofstream out(path,std::ios::app); if(header) out<<"timestamp,scheme,experiment,operation,role,B,trial,warmup,total_time_ns,per_signal_time_ns,scheme_specific_parameters,recipient_mode,status,session_id,session_start_time\n"; char host[256]={};gethostname(host,sizeof(host)-1); auto ts=std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()); const char *sid=std::getenv("OMR_SESSION_ID"); const char *sst=std::getenv("OMR_SESSION_START"); out<<ts<<",OMR,multiple_signal_sender_scalability,clue_generation,sender,"<<B<<","<<trial<<","<<(warm?"true":"false")<<","<<total<<","<<per<<",PVW_n=450;PVW_q=65537;PVW_stddev=1.3;PVW_m=16000;PVW_ell=4,fixed_target_recipient,"<<status<<","<<(sid?sid:"unknown")<<","<<(sst?sst:"unknown")<<"\n";out.flush();
}

int main(int argc, char **argv) {
    if (argc >= 2 && std::string(argv[1]) == "--mode") {
        if (argc != 11 || std::string(argv[2]) != "scaling") { std::cerr << "usage: --mode scaling --signals B --trials N --warmup N --output CSV\n"; return 2; }
        int B=0,trials=0,warmups=0; std::string out; for(int i=3;i<argc;i+=2){std::string k=argv[i],v=argv[i+1];if(k=="--signals")B=std::atoi(v.c_str());else if(k=="--trials")trials=std::atoi(v.c_str());else if(k=="--warmup")warmups=std::atoi(v.c_str());else if(k=="--output")out=v;else return 2;}
        const PVWParam params(450,65537,1.3,16000,4); std::srand(static_cast<unsigned>(std::time(nullptr))); const PVWsk sk=PVWGenerateSecretKey(params); const PVWpk pk=PVWGeneratePublicKey(params,sk); const std::vector<int> zeros(params.ell,0);
        for(int tr=0;tr<warmups+trials;tr++){bool warm=tr<warmups;PVWCiphertext last;auto a=Clock::now();for(int i=0;i<B;i++){PVWCiphertext clue;PVWEncPK(clue,zeros,pk,params);last=clue;}auto b=Clock::now();std::vector<int> decoded;PVWDec(decoded,last,sk,params);bool ok=decoded==zeros;append_scale(out,B,tr-warmups,warm,std::chrono::duration_cast<std::chrono::nanoseconds>(b-a).count(),std::chrono::duration_cast<std::chrono::nanoseconds>(b-a).count()/B,ok?"ok":"failed");if(!ok)return 1;}
        return 0;
    }
    if (argc < 2 || argc > 4) { std::cerr << "usage: omr_sender_smoke CSV [trials] [warmup]\n"; return 2; }
    int trials = argc >= 3 ? std::max(1, std::atoi(argv[2])) : 3;
    int warmups = argc >= 4 ? std::max(0, std::atoi(argv[3])) : 1;
    const PVWParam params(450,65537,1.3,16000,4); std::srand(static_cast<unsigned>(std::time(nullptr)));
    const PVWsk sk=PVWGenerateSecretKey(params); const PVWpk pk=PVWGeneratePublicKey(params,sk);
    const std::vector<int> zeros(params.ell,0);
    for(int trial=0;trial<trials+warmups;trial++) { bool warm=trial<warmups; int measured=trial-warmups;
        PVWCiphertext clue; auto a=Clock::now(); PVWEncPK(clue,zeros,pk,params); auto b=Clock::now();
        std::vector<int> decoded; PVWDec(decoded,clue,sk,params); bool ok=decoded==zeros;
        append(argv[1],measured,warm,"omr_clue_generation",std::chrono::duration_cast<std::chrono::nanoseconds>(b-a).count(),"unclear",ok?"ok":"failed");
        if(!warm) { append(argv[1],measured,false,"omr_sender_preparation",-1,"unclear","unclear"); append(argv[1],measured,false,"message_encryption",-1,"not_applicable","not_applicable"); }
        PVWCiphertext total; auto t0=Clock::now(); PVWEncPK(total,zeros,pk,params); auto t1=Clock::now();
        decoded.clear(); PVWDec(decoded,total,sk,params); bool total_ok=decoded==zeros;
        if (!warm) append(argv[1],measured,false,"omr_sender_total_signaling",std::chrono::duration_cast<std::chrono::nanoseconds>(t1-t0).count(),"unclear",total_ok?"ok":"failed");
        if(!ok||!total_ok) return 1;
    }
    std::cout << "OMR sender smoke ok; server_retrieval=false; file_io=false; output_bytes=unclear\n";
}

File: baselines/pps-gc/pps-garbled-circuits/examples/sender-smoke.rs
use rand::rngs::OsRng;
use rand::Rng;
use std::env;
use std::fs::{metadata, read_to_string, OpenOptions};
use std::io::{Write};
use std::time::{Instant, SystemTime, UNIX_EPOCH};
use stealth_address_circuits::consts::{INDEX_BYTES, LOCATION_BYTES};

fn append(path: &str, trial: usize, warm: bool, op: &str, elapsed: i128,
          bytes: Option<usize>, bytes_status: &str, status: &str) {
    let fresh = metadata(path).map(|m| m.len() == 0).unwrap_or(true);
    let mut f = OpenOptions::new().create(true).append(true).open(path).unwrap();
    if fresh { writeln!(f, "timestamp,scheme,experiment,operation,role,trial,warmup,message_size_bytes,recipient_input_size,scheme_specific_parameters,wall_time_ns,output_bytes,output_bytes_status,status,git_commit,compiler,compiler_version,build_type,machine_id").unwrap(); }
    let ts = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_nanos();
    let host = env::var("HOSTNAME").ok().or_else(|| read_to_string("/etc/hostname").ok()).unwrap_or_else(|| "unknown".into()).trim().to_string();
    writeln!(f, "{},PPS-GC,sender_signaling,{},sender,{},{},32,2,recipient_bits=16;location_bytes=32,{},{},{},{},unavailable,rustc,nightly-2021-05-06,Release,{}",
             ts, op, trial, warm, elapsed, bytes.map(|v|v.to_string()).unwrap_or_default(), bytes_status, status, host).unwrap();
    f.sync_data().unwrap();
}

fn generate(rng: &mut OsRng, recipient: u16, value: &[u8; LOCATION_BYTES]) -> (u16,u16,[u8;LOCATION_BYTES],[u8;LOCATION_BYTES]) {
    let ra: u16 = rng.gen(); let rb = ra ^ recipient;
    let la: [u8; LOCATION_BYTES] = rng.gen(); let mut lb = [0u8; LOCATION_BYTES];
    for i in 0..LOCATION_BYTES { lb[i] = la[i] ^ value[i]; }
    (ra, rb, la, lb)
}

fn prepare(ra:u16, rb:u16, la:&[u8;LOCATION_BYTES], lb:&[u8;LOCATION_BYTES]) -> (Vec<u8>,Vec<u8>) {
    let mut a=Vec::with_capacity(INDEX_BYTES+LOCATION_BYTES); let mut b=Vec::with_capacity(INDEX_BYTES+LOCATION_BYTES);
    a.extend_from_slice(&ra.to_be_bytes()); a.extend_from_slice(la);
    b.extend_from_slice(&rb.to_be_bytes()); b.extend_from_slice(lb); (a,b)
}

fn correct(recipient:u16,value:&[u8;LOCATION_BYTES],ra:u16,rb:u16,la:&[u8;LOCATION_BYTES],lb:&[u8;LOCATION_BYTES])->bool {
    (ra^rb)==recipient && (0..LOCATION_BYTES).all(|i|(la[i]^lb[i])==value[i])
}

fn append_scale(path:&str,b:usize,trial:isize,warm:bool,total:i128,per:i128,status:&str){
 let fresh=metadata(path).map(|m|m.len()==0).unwrap_or(true); let mut f=OpenOptions::new().create(true).append(true).open(path).unwrap();
 if fresh { writeln!(f,"timestamp,scheme,experiment,operation,role,B,trial,warmup,total_time_ns,per_signal_time_ns,scheme_specific_parameters,recipient_mode,status").unwrap(); }
 let ts=SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_nanos(); writeln!(f,"{},PPS-GC,multiple_signal_sender_scalability,share_generation,sender,{},{},{},{},{},recipient_bits=16;location_bytes=32,fixed_target_recipient,{}",ts,b,trial,warm,total,per,status).unwrap(); f.sync_data().unwrap();
}

fn main() {
    let raw: Vec<String> = env::args().collect();
    if raw.get(1).map(|x|x=="--mode").unwrap_or(false) {
        if raw.len()!=11 || raw[2]!="scaling" { panic!("usage: --mode scaling --signals B --trials N --warmup N --output CSV"); }
        let mut b=0usize; let mut trials=0usize; let mut warmups=0usize; let mut path="".to_string(); let mut i=3;
        while i<raw.len(){match raw[i].as_str(){"--signals"=>b=raw[i+1].parse().unwrap(),"--trials"=>trials=raw[i+1].parse().unwrap(),"--warmup"=>warmups=raw[i+1].parse().unwrap(),"--output"=>path=raw[i+1].clone(),_=>panic!("unknown argument")};i+=2;}
        let recipient=7u16; let mut value=[0u8;LOCATION_BYTES]; value.copy_from_slice(b"hi, anonymus signal message here"); let mut rng=OsRng;
        for tr in 0..(trials+warmups){let warm=tr<warmups; let start=Instant::now(); let mut last=(0u16,0u16,[0u8;LOCATION_BYTES],[0u8;LOCATION_BYTES]); let mut sink=0u16; for _ in 0..b {last=generate(&mut rng,recipient,&value); sink ^= last.0 ^ last.1; sink ^= last.2[0] as u16 ^ last.3[0] as u16;} unsafe { std::ptr::read_volatile(&sink); } let total=start.elapsed().as_nanos() as i128; let ok=correct(recipient,&value,last.0,last.1,&last.2,&last.3); append_scale(&path,b,tr as isize-warmups as isize,warm,total,total/b as i128,if ok{"ok"}else{"failed"}); if !ok{std::process::exit(1);}}
        return;
    }
    let mut args=env::args(); let _exe=args.next(); let path=args.next().expect("usage: ppsgc_sender_smoke CSV [trials] [warmup]");
    let trials: usize=args.next().and_then(|x|x.parse().ok()).unwrap_or(3); let warmups: usize=args.next().and_then(|x|x.parse().ok()).unwrap_or(1); let recipient=7u16;
    let mut value=[0u8;LOCATION_BYTES]; value.copy_from_slice(b"hi, anonymus signal message here");
    let mut rng=OsRng; let mut previous_ra=None;
    const GEN_INNER: usize = 100; const PREP_INNER: usize = 1000;
    for trial in 0..(trials+warmups) { let warm=trial<warmups; let measured=trial.saturating_sub(warmups);
        let s=Instant::now(); let mut last=(0u16,0u16,[0u8;LOCATION_BYTES],[0u8;LOCATION_BYTES]);
        for _ in 0..GEN_INNER { last=generate(&mut rng,recipient,&value); }
        let gen=s.elapsed().as_nanos() as i128 / GEN_INNER as i128;
        let (ra,rb,la,lb)=last;
        let p=Instant::now(); let mut a=Vec::new(); let mut b=Vec::new();
        for _ in 0..PREP_INNER { let (aa,bb)=prepare(ra,rb,&la,&lb); a=aa; b=bb; }
        let prep=p.elapsed().as_nanos() as i128 / PREP_INNER as i128;
        let ok=correct(recipient,&value,ra,rb,&la,&lb)&&a.len()==INDEX_BYTES+LOCATION_BYTES&&b.len()==INDEX_BYTES+LOCATION_BYTES&&previous_ra!=Some(ra); previous_ra=Some(ra);
        append(&path,measured,warm,"pps_gc_share_generation",gen,Some(a.len()+b.len()),"measured",if ok{"ok"}else{"failed"});
        append(&path,measured,warm,"pps_gc_sender_preparation",prep,Some(a.len()+b.len()),"measured",if ok{"ok"}else{"failed"});
        if !warm { append(&path,measured,false,"message_encryption",-1,None,"not_applicable","not_applicable"); }
        let t=Instant::now(); let mut ta=Vec::new(); let mut tb=Vec::new(); let mut tra=0; let mut trb=0; let mut tla=[0u8;LOCATION_BYTES]; let mut tlb=[0u8;LOCATION_BYTES];
        for _ in 0..GEN_INNER { let z=generate(&mut rng,recipient,&value); tra=z.0; trb=z.1; tla=z.2; tlb=z.3; let (aa,bb)=prepare(tra,trb,&tla,&tlb); ta=aa; tb=bb; }
        let total=t.elapsed().as_nanos() as i128 / GEN_INNER as i128;
        let tok=correct(recipient,&value,tra,trb,&tla,&tlb)&&ta.len()+tb.len()==2*(INDEX_BYTES+LOCATION_BYTES);
        if !warm { append(&path,measured,false,"pps_gc_sender_total_signaling",total,Some(ta.len()+tb.len()),"measured",if tok{"ok"}else{"failed"}); }
        if !ok||!tok { std::process::exit(1); }
    }
    println!("PPS-GC sender smoke ok; per-server=34 total=68 fresh_randomness=true");
}
